src/Storages/StorageMerge.cpp

```diff
@@ -166,7 +166,7 @@ QueryProcessingStage::Enum StorageMerge::getQueryProcessingStage(
             if (table && table.get() != this)
             {
                 ++selected_table_size;
-                stage_in_source_tables = std::max(
+                stage_in_source_tables = std::min(
```

Comment:
This looks incorrect. Since it's initalized with `FetchColumns`, is always will stay `FetchColumns`. I guess it's the reason why some tests with distributed tables fail.

```
<Error> executeQuery: Code: 8. DB::Exception: Cannot find column `key` in source stream. (THERE_IS_NO_COLUMN) (version 21.9.1.1) (from [::1]:36196) (comment: 01223_dist_on_dist.sql) (in query: select count() from merge_dist_01223;)
```

Without going into details I guess, the following happen. The column `key` is requested only to calculate `count()` as the smallest column. So, distributed table completed query upto `WithMergeStates` and column `key` was dropped, but since the `FetchColumns` stage was returned for `Merge` table, we unsuccessfully tried to find this column in source streams.

Reply:
Yes, I know that this line is the reason for failed tests, but changing this line was an attempt to fix the case when Merge table reads from local and distributed tables at the same time. The problem was that in this case we call read method and it is with stage WithMergeableState but in fact it is not so when it is finished -- it lacked convertion to common structure in this case and we get this exception. https://github.com/ClickHouse/ClickHouse/blob/768378b6d8e79f79ea91de7c4afb0c680135d59e/src/Storages/StorageMerge.cpp#L726.
So changing common stage for all tables from `max` to `min` fixes this case, but breaks another one...